### ISSUE TITLE
`cross_encoder_ranker` + ConvAI2 parameters

### ISSUE THREAD

**Issue by bkj** [facebookresearch/ParlAI#1794]
The [polyencoder paper](https://arxiv.org/pdf/1905.01969.pdf) discuesses a cross-encoder system that gets very high hits@1 on ConvAI2.  I'm trying to reproduce these experiments using the cross-encoder implementation [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/bert_ranker/cross_encoder_ranker.py).

In issue #1713, I was able to (more or less) successfully reproduce the reported performance of the bi-encoder.  Making a couple of modifications described in the paper (different batch size, longer warmup period, using `inline` candidates), this is my best guess at the correct parameterization:

```bash
python -u parlai/scripts/train_model.py                          \
    --task                      convai2                          \
    --model                     bert_ranker/cross_encoder_ranker \
    --dict-file                 $OUTPATH/dictionary              \
    --model-file                $OUTPATH/model                   \
    --shuffle                   true                             \
    --log_every_n_secs          10                               \
    --eval-batchsize            16                               \
    --data-parallel             true                             \
    --type-optimization         all_encoder_layers               \
    --batchsize                 16                               \
    --learningrate              5e-5                             \
    --history-size              20                               \
    --label-truncate            72                               \
    --text-truncate             360                              \
    --lr-scheduler              reduceonplateau                  \
    --lr-scheduler-patience     0                                \
    --lr-scheduler-decay        0.4                              \
    --validation-every-n-epochs 0.5                              \
    --warmup_updates            1000                             \
    --fp16                      true
```

I'm planning to run this on a p3.16xlarge EC2 instance w/ 8xV100 -- but before I do, does anyone see anything wrong w/ the parameters above?  Or (even better) is it possible to share the scripts that were used to run the experiments in the paper?

Thanks!
~ Ben

**Comment by stephenroller** [comment]
when you launch on 8 gpus with dataparallel, you should be able to 8x that batchsize, maybe even more with the fp16 on?

**Comment by bkj** [comment]
The paper says that they use a batch size of 16 but -- there are 19 distractors for every record in ConvAI1, so the number of records being processed per minibatch is 16 * 20 = 320.  Based on my experiments w/ the bi-encoder, I think that's going to get pretty close to the GPU memory limits.

**Comment by stephenroller** [comment]
Oh yes you're right. I get confused with all the models. :)

**Comment by stephenroller** [comment]
Sam and Kurt are both unavailable now, but hopefully one of them can chime in here tomorrow.

**Comment by bkj** [comment]
I ran the above parameters for 3.5 epochs and got the following results
```
{"exs":7801,"accuracy":0.8169,"f1":0.8373,"hits@1":0.817,"hits@5":0.976,"hits@10":0.996,"hits@100":1,"bleu":0.8167,"lr":5e-05,"num_updates":4108,"examples":7801,"loss":5566,"mean_loss":0.7135,"mean_rank":1.436,"mrr":0.8862,"train_time":2709.195165157318}
{"exs":7801,"accuracy":0.8316,"f1":0.8499,"hits@1":0.832,"hits@5":0.979,"hits@10":0.997,"hits@100":1,"bleu":0.8314,"lr":5e-05,"num_updates":8213,"examples":7801,"loss":4560,"mean_loss":0.5845,"mean_rank":1.4,"mrr":0.8951,"train_time":5362.023273229599}
{"exs":7801,"accuracy":0.8312,"f1":0.8495,"hits@1":0.831,"hits@5":0.978,"hits@10":0.996,"hits@100":1,"bleu":0.831,"lr":5e-05,"num_updates":12274,"examples":7801,"loss":5455,"mean_loss":0.6992,"mean_rank":1.397,"mrr":0.8953,"train_time":7982.991955757141}
{"exs":7801,"accuracy":0.8504,"f1":0.8666,"hits@1":0.85,"hits@5":0.983,"hits@10":0.997,"hits@100":1,"bleu":0.8502,"lr":2e-05,"num_updates":16328,"examples":7801,"loss":4707,"mean_loss":0.6034,"mean_rank":1.328,"mrr":0.9088,"train_time":10577.744508981705}
{"exs":7801,"accuracy":0.8575,"f1":0.8732,"hits@1":0.857,"hits@5":0.988,"hits@10":0.997,"hits@100":1,"bleu":0.8574,"lr":8e-06,"num_updates":20395,"examples":7801,"loss":5042,"mean_loss":0.6464,"mean_rank":1.3,"mrr":0.9141,"train_time":13152.8646068573}
{"exs":7801,"accuracy":0.8631,"f1":0.8782,"hits@1":0.863,"hits@5":0.988,"hits@10":0.998,"hits@100":1,"bleu":0.863,"lr":3.2e-06,"num_updates":24497,"examples":7801,"loss":5179,"mean_loss":0.6638,"mean_rank":1.288,"mrr":0.9176,"train_time":15756.786456346512}
{"exs":7801,"accuracy":0.8658,"f1":0.8805,"hits@1":0.866,"hits@5":0.989,"hits@10":0.997,"hits@100":1,"bleu":0.8657,"lr":1.28e-06,"num_updates":28590,"examples":7801,"loss":5179,"mean_loss":0.6639,"mean_rank":1.285,"mrr":0.9189,"train_time":18385.45897293091}
```
the paper reports 87.3 +- 0.3 -- so my results seems about right.

If there's anything Sam and Kurt have to add, I'd be interested to hear, but I'm relatively satisfied with this.

**Comment by samhumeau** [comment]
Hi bkj. The numbers in the paper have been obtained with a slightly different version of that ranker, but the following parameters should give you the same results

```
 python -u examples/train_model.py -pyt convai2 --shuffle true --model bert_ranker/cross_encoder_ranker --batchsize 16 --eval-batchsize 12 --warmup_updates 1000 --lr-scheduler-patience 0 --lr-scheduler-decay 0.4 -lr 5e-05 --data-parallel True --history-size 20 --label-truncate 72 --text-truncate 360 --num-epochs 12.0 --max_train_time 200000 -veps 0.5 -vme 2500 --validation-metric accuracy --validation-metric-mode max --save-after-valid True --type-optimization all_encoder_layers  --log_every_n_secs 20 --candidates inline --fp16 True --model-file ./mycrossencoder45
```

I am going to open source by the end of the week a similar ranker with a model pre-trained on Reddit that scores 90+ on the valid set of convai2

**Comment by stephenroller** [comment]
I'm going to close this issue but anyone can reopen it if there are more action items for the ParlAI team.

**Comment by bkj** [comment]
Great -- thanks everyone!  @samhumeau's parameters look basically the same as mine, w/ a couple of extra things.  Presumably any gaps between my model and the published results are due to training for ~3x longer. 

```
--max_train_time 200000
--num-epochs 12.0 
--validation-metric accuracy 
--validation-metric-mode max 
```

**Comment by samhumeau** [comment]
@bkj Checkout https://github.com/facebookresearch/ParlAI/tree/master/parlai/zoo/pretrained_transformers, 
it shows how to create a crossencoder that scores 90+. 

But we also have a model called polyencoder that is able to score 89+ on convai2, is faster to train (2 hours), and can rank 100k candidates in real time

**Comment by samhumeau** [comment]
(real time is << 1s)

**Comment by bkj** [comment]
Great -- thanks!  I was just about to try to implement the polyencoder model..

**Comment by bkj** [comment]
@samhumeau by the way -- what software did you use to pretrain the transformer models for this?
